src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.OSX.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+
+namespace System.IO
+{
+    internal partial struct FileStatus
+    {
+        internal unsafe void SetCreationTime(string path, DateTimeOffset time)
+        {
+            Interop.Sys.TimeSpec timeSpec = default;
+
+            long seconds = time.ToUnixTimeSeconds();
+            long nanoseconds = UnixTimeSecondsToNanoseconds(time, seconds);
+
+            timeSpec.TvSec = seconds;
+            timeSpec.TvNsec = nanoseconds;
+
+            Interop.libc.AttrList attrList = default;
+            attrList.bitmapCount = Interop.libc.AttrList.ATTR_BIT_MAP_COUNT;
+            attrList.commonAttr = Interop.libc.AttrList.ATTR_CMN_CRTIME;
+
+            // Try to set the attribute on the file system entry using setattrlist,
+            // if we get ENOTSUP then it means that "The volume does not support
+            // setattrlist()", so we fall back to the method used on other unix
+            // platforms, otherwise we throw an error if we get one, or invalidate
+            // the cache if successful because otherwise it has invalid information.
+            Interop.Error result = (Interop.Error)Interop.libc.setattrlist(path, &attrList, &timeSpec, sizeof(Interop.Sys.TimeSpec), default(CULong));
+            if (result == Interop.Error.ENOTSUP)
+            {
+                SetCreationTime_StandardUnixImpl(path, time);
+            }
+            else if (result == Interop.Error.SUCCESS)
+            {
+                InvalidateCaches();
+            }
+            else
+            {
+                Interop.CheckIo(result, path, InitiallyDirectory);
+            }
+        }
+
+        private unsafe void SetAccessOrWriteTime(string path, DateTimeOffset time, bool isAccessTime)
```

Comment:
Why did the access/write time become OSX specific?

Reply:
See https://github.com/dotnet/runtime/blob/f84293583bb0c8e1060bdacfb80158a6c64c5fbd/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.OSX.cs#L55 and the comment below it. It is because `UTimensat` ruins the creation date in some cases.

Reply:
There is no point having this code on other unix platforms as they do not have a proper concept of a creation date, but if they did they would likely suffer a similar issue.

Reply:
I see, on OSX, if you change the write time before the creation time, it will update the creation time to match the write time.

This code compensates for that, by setting the creation time to the original value when that happens.
And OSX at that point is fine with the creation time being past the write time?

It may be an option to accept this as platform specific behavior, and not update the creation time.

What happens when the access time is changed before the creation time?

Does something change when the access time is changed before the write time?

Reply:
Access and write time are always written together (hence the reading of the modification date in the case of setting an access date). There is a test in this PR that checks all the combinations of ordering setting the dates.
The reason that this is even necessary is because the api used to set write and access dates is a base unix api and it ruins creation dates because it only properly considers modification and access dates, if we were to use the bsd api for all of them it would probably not have this issue, but it is a much larger and unnecessary change in my opinion.
Also, if we were to accept this platform behaviour, then we should have a remark on the documentation (which is another reason why we should circumvent it in my opinion).
Also, do note that it would conflict with Windows behaviour if we left it to change the creation date.
Another reason to not leave this is because it is standard behaviour (on both macOS and Windows) when copying a file (with file explorer / finder) is to leave the modification date, and set the creation date to the time at which it was copied. Without doing this, someone emulating this behaviour would have to make sure they set the dates in the correct order, but they would probably not be aware that the order matters, with this change it avoids that situation.

Reply:
I don't think most users would expect that setting the modification date can change the creation date.

Reply:
I agree, there is value in having this compatibility with Windows and let users set creation time and write time in any order.

Reply:
I personally ran into this issue (I didn't just accidentally discover it reading documentation), which is why I decided to fix it on here instead of reporting an issue. I'd imagine if someone did run into the issue they would look at the documentation and discover that you can't set the date on macOS and think 'oh, that's just how it is then since it's not supported on macOS', not knowing that it's an issue with the apis .NET uses.

Reply:
I also think we should improve the Windows compat.